x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+
+  return {
+    immutable,
+    rule_source,
+    // [... more object properties ...]
+  };
+};
+```
+
+And the `normalizeRuleSourceSchemaOnRuleRead` can be defined so:
+
+_Source: x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/prebuilt_rule_schema_migrations.ts_ (New file)
+
+```ts
+interface OnReadNormalizationResult {
+  immutable: IsRuleImmutable;
+  rule_source: RuleSource
+}
+
+/**
+ * To calculate `rule_source`:
+ * - Use `rule_source` field  if already exists in the internal rule object params.
+ * - If it does not exist, check the value `immutable` from the rules param:
+ *   - if it is `true`, create a new `rule_source` object, with `isCustomized` set to `false` and `repoNam` set to `elastic_prebuilt` (Use case of rule_source rules that have not been yet migrated-on-write)
+ *   - otherwise, the field should be `undefined` (use case of custom rules).
+ */
+const getRuleSourceValueForRuleRead = (ruleParams: BaseRuleParams): RuleSource | undefined => {
+  if (ruleParams.rule_source) {
+    return ruleParams.rule_source;
+  }
+
+  if (ruleParams.immutable) {
+    return {
+      type: 'external'
+      isCustomized: false,
+    };
+  }
+
+  return {
+    type: 'internal'
+  };
+};
+
+export const normalizePrebuiltSchemaOnRuleRead = (
+  ruleParams: BaseRuleParams
+): OnReadNormalizationResult => {
+
+/**
+ * To calculate `immutable`:
+ * - Checks if the `rule_source` field exists in the rule's parameters.
+ *   - If `rule_source` exists and its type is external, sets `immutable` to `true`. (Use case of Rules that have already been migrated-on-write)
+ *   - If `rule_source` does not exist, return the value of the params' `immutable` field. (Use case of Rules that have not yet been migrated on write.)
+ */
+  const isRuleSourceExternal = ruleParams.rule_source && ruleParams.rule_source.type === 'external';
+  const immutable = Boolean(isRuleSourceExternal) || Boolean(ruleParams.immutable);
+  const rule_source = getRuleSourceValueForRuleRead(ruleParams);
+
+  return {
+    immutable,
+    rule_source,
+  };
+};
+```
+
+##### Rule Management endpoints that will perform normalization-on-read
+
+- All endpoints that also perform migration-on-write, as listed in the below `Migration on write` section. All of them use the `internalRuleToAPIResponse` methods to transform the rule before returning the response, so the normalization on read will take place there. For these specific cases, however, the normalization-on-read will take place, but the data stored in ES for each rule should match the response given by the endpoint, i.e. the normalized in-memory value and the stored value of the rule should be the same.
+- **Find Rules - `GET /rules/_find`**
+- **Read Rule - GET /rules:**
+- **Delete Rules - DELETE /rules**
+- **Bulk Actions** - `POST /rules/_bulk_action`: all other action types not mentioned in the previous section will perform normalization-on-read before responding:
```

Comment:
What previous section do you refer here? Did I get it right only `Enable`, `Disable`, `Delete` and `Export` perform migration-on-write?

Reply:
What's special about the Enable, Disable, Delete and Export bulk actions? Shouldn't just all actions perform normalization-on-read?

Reply:
Yes, sorry, this is bad redaction. All bulk action types will perform normalization on read. The list here tried to explain that Enable, Disable, Delete and Export ONLY do normalization-on-read, and NOT migration-on-write.

I'll replace all of this with a table anyways.

Reply:
@jpdjere The table looks much better 👍